### PR TITLE
User without tax number erica

### PR DESCRIPTION
**Pull request by JulianRoesner**
# Short Description
Adapt Erica to enable it to allow correct processing of tax declaration without a tax number.

# Changes
- Erica input: Extension by a boolean flag (new_admission), which must be set if no tax number is specified, and the bufa_nr.
- Change generate_vorsatz, generate_full_est_xml and EstValidationRequestController to set the correct values in the xml if the flag is set
- Allow the MockErica to also allow a tax declaration without a tax number

# Feedback
- Do you see any missing tests?
- I'll highlight some parts in the code.

### CODE REVIEW

File: erica_app/erica/routes.py [hunk 1]
```diff
@@ -10,10 +10,10 @@
     UnlockCodeActivationRequestController, EstValidationRequestController, EstRequestController, \
     UnlockCodeRevocationRequestController
 
-PYERIC_VERSION_URL = '/01'
+ERICA_VERSION_URL = '/01'
```
Comment (JulianRoesner) [inline erica_app/erica/routes.py:13]: I would propose to update this version either now or when we add a new route for getting the list of the tax offices. WDYT?

File: erica_app/tests/elster_xml/test_elster_xml_generator.py [hunk 2]
```diff
@@ -710,9 +706,109 @@ def test_add_nutzdaten_header(self):
         self.assertIn("<NutzdatenTicket>nutzdatenTicket123</NutzdatenTicket>", xml_string)
         self.assertIn('<Empfaenger id="F">9198</Empfaenger>', xml_string)
 
+
+class TestGenerateFullEstXML(unittest.TestCase):
```
Comment (JulianRoesner) [inline erica_app/tests/elster_xml/test_elster_xml_generator.py:710]: I've split some of these tests into to separate classes because there was quite some stuff mixed together

File: erica_app/erica/routes.py (hunk 1 above)
Comment (nfelger) [reply erica_app/erica/routes.py:13]: Hm, I disagree. I would suggest we upgrade the version no sooner than when the following criteria are met:
- we introduce a breaking change
- we have either have clients we do not control (public or other teams) OR we cannot upgrade all clients at the same time as we change Erica.

As it stands, we have a single client under our full control, which we can adapt in the same PR, so changing the version would introduce work with no gain.

Comment (JulianRoesner) [reply erica_app/erica/routes.py:13]: Okay, point taken.

File: erica_app/erica/elster_xml/elster_xml_generator.py [hunk 3]
```diff
@@ -109,7 +109,7 @@ def generate_full_xml(th_fields, nutzdaten_header_generator, nutzdaten_generator
 
 
 def generate_full_est_xml(form_data, steuernummer, year, person_a_idnr, first_name, last_name, street, street_nr, plz,
-                          town, empfaenger, nutzdaten_ticket="1", th_fields=None, use_testmerker=False,
+                          town, empfaenger, nutzdaten_ticket="1", new_admission=False, th_fields=None, use_testmerker=False,
```
Comment (nfelger) [inline erica_app/erica/elster_xml/elster_xml_generator.py:114]: Adding all these boolean flags one by one is a bit of a code smell to me: they introduce a lot of tangled/coupled code, which can be hard to maintain (see https://martinfowler.com/bliki/FlagArgument.html for a more eloquent problem description).

I don't see a good way yet to split this into separate functions, do you? Maybe a small step could be to extract a `_add_tax_nr(vorsatz, new_admission)` function from `_generate_vorsatz` to be called on line 127?

**Review by nfelger — APPROVED** [review]
Looks good to me! Left some comments.

File: erica_app/erica/request_processing/erica_input.py [hunk 4]
```diff
@@ -78,12 +80,28 @@ class FormDataEst(BaseModel):
...
     def must_be_correct_length(cls, v):
-        if not 10 <= len(v) <= 11:
+        if v and not 10 <= len(v) <= 11:
             raise ValueError('must be 10 or 11 numbers long')
         return v
 
+
+
+    @validator('bufa_nr', always=True)
+    def if_new_admission_bufa_nr_must_be_set_correctly(cls, v, values):
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply erica_app/erica/request_processing/erica_input.py:100]: Uh, that would be nice. Currently, we only validate that through Elster. However, adding this here the error would then be part of a validation error instead of a wrong bufa-number error. 

Another solution would be to check it in there the request processing.

File: erica_app/erica/request_processing/erica_input.py [hunk 5]
```diff
@@ -6,7 +6,9 @@
 
 
 class FormDataEst(BaseModel):
-    steuernummer: str
+    steuernummer: Optional[str]
+    new_admission: Optional[bool]
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply erica_app/erica/request_processing/erica_input.py:10]: Yes, I was also thinking about sticking to the German name "Neuaufnahme". Personally, I would go with `submission_without_tax_nr ` because `request_tax_nr ` sounds as if we will send a tax_nr back from Erica

File: erica_app/erica/elster_xml/elster_xml_generator.py (hunk 3 above)
Comment (JulianRoesner) [reply erica_app/erica/elster_xml/elster_xml_generator.py:114]: Yes, I've added the flags only because I wanted to make it explicit. 

So, basically something like:
```python
     vorsatz = _generate_vorsatz(steuernummer, year, person_a_idnr, person_b_idnr, first_name, last_name, street,
                                 street_nr, plz, town)
                                 street_nr, plz, town)
     if new_admission:
          _add_tax_nr(vorsatz, new_admission) 
     _add_xml_nutzdaten_header(nutzdaten_block_xml, nutzdaten_ticket, empfaenger)
```

The other way would be to take `generate_vorsatz` out of this function and have vorsatz as parameter. We do not need the xml for the vorsatz. So, this could happen anywhere else. And then we could have two functions for vorsatz: `generate_vorsatz_with_tax_nr` and `generate_vorsatz_without_tax_nr`

File: erica_app/erica/request_processing/erica_input.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (nfelger) [reply erica_app/erica/request_processing/erica_input.py:10]: Hm, a tricky line to toe there: "Neuaufnahme" is probably what the documentation says, but I _also_ wouldn't really know at a glance that this means "this person has no tax nr". (It's different for technical terms like "bufa_nr" that are unambiguously in the tax domain.)

In this case, I'd go for a name that is clear to us, even if it doesn't match the German documentation.

[the remark this answers is not in the document]
Comment (nfelger) [reply erica_app/erica/request_processing/erica_input.py:10]: Let's use `submission_without_tax_nr`

File: erica_app/erica/request_processing/erica_input.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (nfelger) [reply erica_app/erica/request_processing/erica_input.py:100]: Ok, I don't know enough about the wiring of validations in Erica to make any meaningful suggestion, I think.

File: erica_app/erica/elster_xml/elster_xml_generator.py (hunk 3 above)
Comment (nfelger) [reply erica_app/erica/elster_xml/elster_xml_generator.py:114]: I like your second suggestion!

File: erica_app/erica/request_processing/erica_input.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply erica_app/erica/request_processing/erica_input.py:100]: If we raise an error during in the pydantic validator, fastapi will send its own error message and wrap the InvalidBufaNumber error inside that message.

If we would check the BufaNumber later, we could raise our own InvalidBufaNumber error and also send that to the webapp. That said, we are currently not excepting these kind of errors and if we will, we can use the validation error, too. And I think it makes more sense to define it here and ensure that only valid bufa numbers get through. Therefore I've added a check to the validator.

TLDR: I've implemented your idea 😉

[the remark this answers is not in the document]
Comment (nfelger) [reply erica_app/erica/request_processing/erica_input.py]: FWIW you can [override the fastapi exception handler that generates the validation error response](https://fastapi.tiangolo.com/tutorial/handling-errors/#override-request-validation-exceptions). (Although I think we should try to work with the defaults unless we have a compelling reason not to.)

File: erica_app/erica/request_processing/requests_controller.py [hunk 6]
```diff
@@ -88,29 +88,41 @@ def process(self):
...
+                electronic_steuernummer,
+                self.input_data.meta_data.year,
+                self.input_data.est_data.person_a_idnr,
+                self.input_data.est_data.person_b_idnr,
+                self.input_data.est_data.person_a_first_name,
+                self.input_data.est_data.person_a_last_name,
+                self.input_data.est_data.person_a_street,
+                self.input_data.est_data.person_a_street_number,
+                self.input_data.est_data.person_a_plz,
+                self.input_data.est_data.person_a_town)
```
Comment (nfelger) [inline erica_app/erica/request_processing/requests_controller.py]: We could make it more explicit what's shared and what isn't like this:

```python
        common_vorsatz_args = (
                self.input_data.meta_data.year,
                self.input_data.est_data.person_a_idnr,
                self.input_data.est_data.person_b_idnr,
                self.input_data.est_data.person_a_first_name,
                self.input_data.est_data.person_a_last_name,
                self.input_data.est_data.person_a_street,
                self.input_data.est_data.person_a_street_number,
                self.input_data.est_data.person_a_plz,
                self.input_data.est_data.person_a_town
        )
        if self.input_data.est_data.submission_without_tax_nr:
            empfaenger = self.input_data.est_data.bufa_nr
            vorsatz = generate_vorsatz_without_tax_number(*common_vorsatz_args)
        else:
            electronic_steuernummer = est_mapping.generate_electronic_steuernummer(
                self.input_data.est_data.steuernummer,
                self.input_data.est_data.bundesland,
                use_testmerker=self._is_testmerker_used())
            vorsatz = generate_vorsatz_with_tax_number(electronic_steuernummer, *common_vorsatz_args)
```

WDYT? Makes it pretty clear that the only difference is that one gets the electro-tax-nr, the other doesn't.

Comment (JulianRoesner) [reply erica_app/erica/request_processing/requests_controller.py]: Great idea!
